app/code/Magento/Catalog/view/frontend/layout/catalog_category_view.xml

```diff
@@ -42,12 +42,15 @@
                 </arguments>
             </block>
         </referenceContainer>
-        <referenceBlock name="page.main.title">
-            <arguments>
-                <argument name="id" xsi:type="string">page-title-heading</argument>
-                <argument name="add_base_attribute_aria" xsi:type="string">page-title-heading toolbar-amount</argument>
-            </arguments>
-            <block class="Magento\Catalog\Block\Category\Rss\Link" name="rss.link" template="Magento_Catalog::category/rss.phtml"/>
-        </referenceBlock>
+        <referenceBlock name="page.main.title" remove="true"/>
```

Comment:
I believe I read somewhere that changing block name is backward-incompatible change but http://devdocs.magento.com/guides/v2.2/contributor-guide/backward-compatible-development/index.html#magento-apis is quite vague about it.

Can we preserve `page.main.title` name but just change block type on needed places?

Reply:
I have thought the same thing, I had problems in the past removing blocks and trying to re-add them with the same name. I'm trying now to override block class definition by setting the same block name, so full block node gets overriden during merge, I keep you updated.

Reply:
Probably it can be done via reference: http://devdocs.magento.com/guides/v2.0/frontend-dev-guide/layouts/xml-manage.html#set_template

Reply:
I think this is for defining / changing arguments for a given block, like the template, but for a given (already instantiated) block. I've achieved changing block without changing name redefining the block, see https://github.com/magento/magento2/pull/11709/commits/584cfd33f3ab7a0bab86e2f303a28c1c95c477da

Reply:
Good! But, will replacing `class` using `referenceBlock` work? It would allow to not add additional `referenceContainer` and generally referencing is safer that redeclaring.

Reply:
I don't think this is possible without rewriting xml node block redefining it with the same name, I've been investigating this possibility, and this is what I found.

Here is the xsd definition for `referenceBlock`, from `Magento/Framework/View/Layout/etc/elements.xsd`:
```
<xs:complexType name="blockReferenceType" mixed="true">
        <xs:choice minOccurs="0" maxOccurs="unbounded">
            <xs:element name="action" type="actionType" minOccurs="0" maxOccurs="unbounded">
                <xs:annotation>
                    <xs:documentation>
                        Argument name must be unique in scope of action.
                    </xs:documentation>
                </xs:annotation>
                <xs:key name="blockReferenceActionArgumentName">
                    <xs:selector xpath="argument"></xs:selector>
                    <xs:field xpath="@name"></xs:field>
                </xs:key>
            </xs:element>
            <xs:element ref="arguments" minOccurs="0" maxOccurs="1"/>
            <xs:element ref="block" minOccurs="0"/>
            <xs:element name="container" type="containerType" minOccurs="0"/>
            <xs:element ref="referenceBlock" minOccurs="0" />
            <xs:element ref="referenceContainer" minOccurs="0"/>
            <xs:element ref="uiComponent" minOccurs="0" />
        </xs:choice>
        <xs:attribute type="elementNameType" name="name" use="required"/>
        <xs:attribute type="xs:string" name="template" use="optional"/>
        <xs:attribute type="xs:boolean" name="display" default="true" use="optional"/>
        <xs:attribute type="xs:boolean" name="remove" use="optional"/>
    </xs:complexType>
```

To change the class, I think it should be done redefining class as a `referenceBlock` attribute: ```<referenceBlock name="page.main.title" class="Magento\Theme\Block\Html\RawTitle">```, thing that is not allowed by xsd definition (it only allows name, template, display and remove attributes for `referenceBlock`).

The rest of the xsd declaration are arguments for the already defined block class, or actions over it, or declaring children blocks, or referencing other blocks, containers or uiComponents.

I have even tried to find any example of any layout xml file trying to perform an action like this with no success. No clue of this kind of usage even in layout directives tests under Magento/Framework/View/_files/layout_directives_test.

If you still don't like the idea of doing this task the way it is done, I could still use `$this->eventManager->dispatch('layout_render_before_' . $this->request->getFullActionName());` event placed in `\Magento\Framework\View\Result\Layout::renderResult`, or tweak it in some way in `\Magento\Catalog\Block\Category\View::_prepareLayout` and `\Magento\Catalog\Block\Product\View::_prepareLayout`, but this way I think you may lose some control about this, control that you may have per theme editing layout xml files.

Reply:
> thing that is not allowed by xsd definition

Thanks, that's pretty answers my question. Current implementation with declarative approach instead of some PHP code seems best to me. Will check if there was some intentional decision to disallow class overriding.

Did class overwriting work properly by the way if you expand XSD?

Reply:
Surprisingly, yes, it does, this way:
- At Magento/Framework/View/Layout/etc/elements.xsd:
<img width="662" alt="captura de pantalla 2017-10-26 a las 17 03 36" src="https://user-images.githubusercontent.com/17545750/32060705-e0d16a34-ba6f-11e7-845d-a9994d319786.png">
- At Magento/Catalog/view/frontend/layout/catalog_product_view.xml:
<img width="977" alt="captura de pantalla 2017-10-26 a las 17 07 11" src="https://user-images.githubusercontent.com/17545750/32060828-38d85602-ba70-11e7-9287-ae0a41728111.png">
- Result:
![captura de pantalla 2017-10-26 a las 17 08 25](https://user-images.githubusercontent.com/17545750/32060859-5003b8bc-ba70-11e7-89df-bd8cf16ace8e.png)

- Only removing class attribute from referenceBlock:
<img width="970" alt="captura de pantalla 2017-10-26 a las 17 09 22" src="https://user-images.githubusercontent.com/17545750/32060906-73fe1b9a-ba70-11e7-87bb-5666fe3f780d.png">
- Result:
![captura de pantalla 2017-10-26 a las 17 09 58](https://user-images.githubusercontent.com/17545750/32060935-8e91ca1a-ba70-11e7-851f-d767db91a0e2.png)

So yes, it works. Have you found any intentional decision for this not being allowed by xsd?

Reply:
Nah, didn't ask yet. Asking now. Big thanks for checking!